consensus/hotstuff/signature/packer.go

```diff
@@ -229,3 +168,91 @@ func deserializeFromBitVector(serialized []byte, count int) ([]hotstuff.SigType,
 
 	return types, nil
 }
+
+// encodeSignerIndicesAndSigType encodes the given stakingSigners and beaconSigners into signer indices and sig types.
+// The given fullMembers provides the canonical order of signer index for each signer.
+// For instance, assuming fullMembers are [A,B,C,D,E,F], stakingSigners are [D,E], beaconSigners are [A,B], then
+// the signerIndices will be [byte([1,1,0,1,1,0,0,0])],
+// 			bit 1 indicates the node at that index signed, bit 0 indicates the node at that index didn't sign
+// the sigType will be [byte([1,1,0,0,0,0,0,0])],
+// 			bit 1 indicates the signer at the same index in signerIndices signed random beacon sig
+// 			bit 0 indicates the signer at the same index in signerIndices signed staking sig
+func encodeSignerIndicesAndSigType(fullMembers []flow.Identifier, stakingSigners flow.IdentifierList, beaconSigners flow.IdentifierList) ([]byte, []byte, error) {
+	stakingSignersLookup := stakingSigners.Lookup()
+	beaconSignersLookup := beaconSigners.Lookup()
+
+	indices := make([]int, 0, len(fullMembers))
+	sigType := make([]hotstuff.SigType, 0, len(fullMembers))
+
+	for i, member := range fullMembers {
+		if _, ok := stakingSignersLookup[member]; ok {
+			indices = append(indices, i)
+			delete(stakingSignersLookup, member)
+
+			sigType = append(sigType, hotstuff.SigTypeStaking)
+			continue
+		}
+
+		if _, ok := beaconSignersLookup[member]; ok {
+			indices = append(indices, i)
+			delete(beaconSignersLookup, member)
+
+			sigType = append(sigType, hotstuff.SigTypeRandomBeacon)
+			continue
+		}
+	}
+
+	if len(stakingSignersLookup) > 0 {
+		return nil, nil, fmt.Errorf("unknown staking signers %v", stakingSignersLookup)
+	}
+
+	if len(beaconSignersLookup) > 0 {
+		return nil, nil, fmt.Errorf("unknown beacon signers %v", beaconSignersLookup)
+	}
```

Comment:
There are other illegal inputs that could trigger these sanity checks. For example, a node being listed in both `stakingSigners` _and_ `beaconSigners` constitutes an illegal input, which would trigger the latter check. I think we could consolidate both checks and write:

```suggestion
	if len(stakingSignersLookup) > 0 {
		return nil, nil, fmt.Errorf("unknown staking signers %v", stakingSignersLookup)
	}
	if len(beaconSignersLookup) > 0 {
		return nil, nil, fmt.Errorf("unknown or duplicated beacon signers %v", beaconSignersLookup)
	}
```